### ISSUE TITLE
fmt shared build lacking symbols

### ISSUE THREAD

**Issue by stvriver** [fmtlib/fmt#2011]
Hello,

I configured the latest master fmt to build with shared mode. The build processes successfully.
Then I build local_test with #define FMT_HEADER_ONLY no problem.
But the build failed without FMT_HEADER_ONLY, the errors are:

2>    locale-test.obj : error LNK2019: unresolved external symbol "struct fmt::v7::detail::dragonbox::decimal_fp<float> __cdecl fmt::v7::detail::dragonbox::to_decimal<float>(float)" (??$to_decimal@M@dragonbox@detail@v7@fmt@@YA?AU?$decimal_fp@M@0123@M@Z) referenced in function "class fmt::v7::detail::buffer_appender<wchar_t> __cdecl fmt::v7::detail::write<wchar_t,class fmt::v7::detail::buffer_appender<wchar_t>,float,0>(class fmt::v7::detail::buffer_appender<wchar_t>,float)" (??$write@_WV?$buffer_appender@_W@detail@v7@fmt@@M$0A@@detail@v7@fmt@@YA?AV?$buffer_appender@_W@012@V3012@M@Z)
2>    locale-test.obj : error LNK2019: unresolved external symbol "struct fmt::v7::detail::dragonbox::decimal_fp<double> __cdecl fmt::v7::detail::dragonbox::to_decimal<double>(double)" (??$to_decimal@N@dragonbox@detail@v7@fmt@@YA?AU?$decimal_fp@N@0123@N@Z) referenced in function "class fmt::v7::detail::buffer_appender<wchar_t> __cdecl fmt::v7::detail::write<wchar_t,class fmt::v7::detail::buffer_appender<wchar_t>,double,0>(class fmt::v7::detail::buffer_appender<wchar_t>,double)" (??$write@_WV?$buffer_appender@_W@detail@v7@fmt@@N$0A@@detail@v7@fmt@@YA?AV?$buffer_appender@_W@012@V3012@N@Z)

Edit:
I found that the above symbols are required by other tests also, such as chrono-test, color-test, format-test, etc.
Exception: os-test gives this message:
 os-test.obj : error LNK2001: unresolved external symbol "private: virtual void __cdecl fmt::v7::ostream::grow(unsigned __int64)" (?grow@ostream@v7@fmt@@EEAAX_K@Z)
1>    D:\vcpkg\buildtrees\fmt\x64-windows\bin\Debug\os-test.exe : fatal error LNK1120: 1 unresolved externals

Please check.
I have intention to write to https://github.com/fmtlib/fmt/issues/1967 but it was closed.

**Comment by vitaut** [comment]
Fixed in https://github.com/fmtlib/fmt/commit/6d14f7811516ff1a88aff6edbd8c46e607aa7f07, thanks for reporting.

**Comment by stvriver** [comment]
I have found another issue not about missing symbols but redundant symbols.
If you put the function below into more than one cpp file, the issue should occur.

```
#include "fmt/format.h"
void test_memory_buffer() {
	fmt::memory_buffer buff;
	buff.push_back('A');
}

```
```

1>fmtd.lib(fmtd.dll) : error LNK2005: "protected: __cdecl fmt::v7::detail::buffer<char>::buffer<char>(char *,unsigned __int64,unsigned __int64)" (??0?$buffer@D@detail@v7@fmt@@IEAA@PEAD_K1@Z) already defined in second_src_file.obj
1>fmtd.lib(fmtd.dll) : error LNK2005: "protected: void __cdecl fmt::v7::detail::buffer<char>::set(char *,unsigned __int64)" (?set@?$buffer@D@detail@v7@fmt@@IEAAXPEAD_K@Z) already defined in second_src_file.obj
1>fmtd.lib(fmtd.dll) : error LNK2005: "public: unsigned __int64 __cdecl fmt::v7::detail::buffer<char>::size(void)const " (?size@?$buffer@D@detail@v7@fmt@@QEBA_KXZ) already defined in second_src_file.obj
1>fmtd.lib(fmtd.dll) : error LNK2005: "public: unsigned __int64 __cdecl fmt::v7::detail::buffer<char>::capacity(void)const " (?capacity@?$buffer@D@detail@v7@fmt@@QEBA_KXZ) already defined in second_src_file.obj
1>fmtd.lib(fmtd.dll) : error LNK2005: "public: char * __cdecl fmt::v7::detail::buffer<char>::data(void)" (?data@?$buffer@D@detail@v7@fmt@@QEAAPEADXZ) already defined in second_src_file.obj
1>fmtd.lib(fmtd.dll) : error LNK2005: "public: void __cdecl fmt::v7::detail::buffer<char>::push_back(char const &)" (?push_back@?$buffer@D@detail@v7@fmt@@QEAAXAEBD@Z) already defined in second_src_file.obj
```

**Comment by vitaut** [comment]
`buffer` ctor is inline so I'm not sure why you get this error. Are you using CMake?

**Comment by stvriver** [comment]
Yes I am using CMake to generate MSVC project files. 
Why do you not make a trivial step as I mentioned previously? Then you have the clue easily.

**Comment by stvriver** [comment]
I found the cause of redundant symbols. It was introduced in your commit [6d14f78](https://github.com/fmtlib/fmt/commit/6d14f7811516ff1a88aff6edbd8c46e607aa7f07)
```
class FMT_API ostream final : private detail::buffer<char> {
}
```
Removing FMT_API from the declaration in os.h and "grow()" in os.cpp will solved the issue. 
In fact, all functions of class ostream are inline except function "void grow(size_t)" but it is never invoked outside scope of fmt so no need to mark export also.

**Comment by vitaut** [comment]
Thanks for investigating. The issue is tentatively fixed in https://github.com/fmtlib/fmt/commit/5555651ce02a28db8941c49ad60a8a4842777691.

**Comment by haquocviet** [comment]
> Thanks for investigating. The issue is tentatively fixed in [5555651](https://github.com/fmtlib/fmt/commit/5555651ce02a28db8941c49ad60a8a4842777691).

I have just upgraded fmt-7x to fmt-8.x and found that the "redundant symbols" bug is appeared again. 
Please check the issue.
